### ISSUE TITLE
Method param default values : Feature Request

### ISSUE THREAD

**Issue by brandonseydel** [dotnet/roslyn#11143]
Can we add some more express ways to define defaul values for method params?

(string emailAddress = emailAddress?.ToLower()) //email address always lower if a value is passed

**Comment by vcsjones** [comment]
Optional parameters today cannot do that because they must be known at compile time. This isn't an arbitrary constraint. The compiler must know the default value at compile time because the compiler emits calls to the method by simply taking the default value of the parameter and using that as the parameter's value.

A related feature request might be a new syntax for optional parameters whose values are determined at run time, but that's fairly problematic in terms of how they are exposed outside of the assembly. All of that said, this works just as well:

```
public void Foo(string emailAddress)
{
    emailAddress = emailAddress?.ToLower();
}
```

Your suggestion just moves the expression from the method body into the parameter.

**Comment by brandonseydel** [comment]
You can write everything in for loops also.  That isn't the point.  It simplifies the code while maintaining legibility.  I don't get how it is problematic at all for exposure because everyone calling in will only know it as a string.

**Comment by brandonseydel** [comment]
Why would you reference a feature request without giving a link to it?  Please include the reference in your comment so others can see.

**Comment by dsaf** [comment]
It's not a default value. It's an inline expression which simultaneously proposes a new non-lambda syntax for expressions.

**Comment by vcsjones** [comment]
>It simplifies the code while maintaining legibility.

I disagree. Like I said, you just moved code from one place to another when it would work verbatim in the method body. Meanwhile, there is a lot to consider here. How would overriding a method that has one of these parameters work? What if it's abstract or on an interface?

**Comment by dsaf** [comment]
There is another way to look at this. Could it be about declaring a method contract that a string parameter value has to be in lowercase?

https://github.com/dotnet/roslyn/issues/119

```csharp
public void Sample(string emailAddress)
    requires emailAddress == emailAddress?.ToLower()
{
//...
}
```

Of course this is not exactly same because the original sample *guarantees* lower case while mine fails if it's not lowercase. So the proposal could be:

```csharp
public void Sample(string emailAddress)
    prepares emailAddress = emailAddress?.ToLower()
{
//...
}
```

So it's a like a "pre-action" instead of a precondition and piggy-backs a planned flexible feature.

**Comment by brandonseydel** [comment]
Different way of thinking about it also 
```
public void Test(string emailAddress, bool? dumbo) : emailAddress.ToLower(), dumbo ?? false
{
return emailAddress;
}
```

**Comment by alrz** [comment]
Or,
```cs
public void Test(string emailAddress) {
         emailAddress = emailAddress.ToLower();
}
```

**Comment by DavidArno** [comment]
Having thought about this suggestion, I've now removed my down-vote and have applied it to @alrz's comment instead. To explain this, it's a case of how complex the code is to follow in your head. With @alrz's example, it's a case of:
```cs
public void Test(string emailAddress) // emailAddress has a value
{
    emailAddress = emailAddress?.ToLower() ?? "string is empty"; // email address now has
                                                                 //  a new value...
}
```
The re-assignment of `emailAddress` adds to the cognitive complexity of the code. Whenever a value of a variable changes, I have to hold that fact in my head whilst reading the code. 10 lines later, that can become a problem; if I'm faced with a 500 line method, with dozens of ever-changing variables, it becomes a nightmare.

However, I don't think that @brandonseydel's solution is right either. Whilst the original example changes the value of `emailAddress` in the same line as declaring it (and thus is easier to follow mentally), it's still uses the "mutate variables" approach. The solution I'd take (which the language already supports, and which can be enforced with analyzers), is to declare a new variable to take the mutated value:
```cs
public void Test(string rawEmailAddress) 
{
    var correctedEmailAddress = rawEmailAddress.ToLower();
    ...
}
```

(I deliberately also removed the `?` as I don't like defensive code; I prefer the TDD approach of having a test fail is I ended up passing a `null` into `Test`.

**Comment by brandonseydel** [comment]
I don't get why people are just putting stuff that can already be done.  Everyone knows it can be done that way.  The point is to enhance the capabilities of the language not to post asshole responses.  If you don't like it then explain why.  Nobody needs your primitive examples of how it can be done a different way. *cough cough* @alrz @vcsjones

**Comment by alrz** [comment]
I didn't suggest alternatives, I just thought about how the compiler could translate this code. Furthermore, if I could write the whole program inside method's parameter list then it would worth it and makes it really concise compared to the existing method syntax which I can't even.

**Comment by svick** [comment]
@brandonseydel 

> I don't get why people are just putting stuff that can already be done.

Because comparing proposed language feature against what can be done in the existing language is very useful when evaluating how useful a feature is.

> The point is to enhance the capabilities of the language

That's not the only metric that's important. Things like simplicity of language and not being surprising (`string emailAddress = null` does one thing, but `string emailAddress = emailAddress?.ToLower()` does something very different?) are important too.

>  not to post asshole responses

Please don't call people names, it doesn't help anybody.

>  If you don't like it then explain why.

1. That's not how it works. If you do like it, you have to make the case why this would be a useful feature, since [features start at -100 points](https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/).
2. I think that @vcsjones did explain why he didn't like it. And @alrz didn't say he didn't like it.

>  Nobody needs your primitive examples of how it can be done a different way.

I disagree, if you don't evaluate alternatives (both potential and existing), you're not going to find the best choice.
